src/DynamoCore/Scheduler/AsyncTaskExtensions.cs

```diff
@@ -27,7 +27,7 @@ public static IDisposable Then(this AsyncTask task, AsyncTaskCompletedHandler ac
         /// <returns>An IDisposable representing the event subscription</returns>
         public static IDisposable ThenPost(this AsyncTask task, AsyncTaskCompletedHandler action, SynchronizationContext context = null)
         {
-            if (context == null) context = new SynchronizationContext(); // uses the default
+            if (context == null) context = SynchronizationContext.Current ?? new SynchronizationContext(); // uses the current or a default
```

Comment:
SynchronizationContext.Current will be the DIspatcherSyncCOntext in Dynamo + UI.
SynchronizationContext.Current will probably be null or the default sync context  SynchronizationContext in headless mode.

Reply:
Is the `SynchronizationContext` returned here using `SynchronizationContext.Current` the one set on the scheduler thread?

Reply:
SynchronizationContext.Current will be the the **current thread**
s SyncContext (which can be set using SynchronizationContext.SetSynchronizationContext)

If we call `ThenPost` from the UI thread, the the SynchronizationContext.Current will be the dispatcher sync context.
If we call `ThenPost` from a non-UI thread (ex. scheduler thread) the the SynchronizationContext.Current will either be null or whatever sync context was set using the SynchronizationContext.SetSynchronizationContext API.

From what I can tell, the main use cases for the `ThenPost` APIs (at the moment) are to run async tasks that interact with the UI, after some scheduler task has been completed. 

There was a small risk that someone might call `ThenPost` with the default arguments (null context) from a non-UI thread, with the idea to run the `AsyncTaskCompletedHandler`  to interact with the UI. If we were to use the `new SynchronizationContext()`, then that would call the `AsyncTaskCompletedHandler ` in the same thread that called `ThenPost` (in this example, a non UI thread).

Reply:
```
SyncronizationContext:
        public virtual void Post(SendOrPostCallback d, object? state)
            => ThreadPool.QueueUserWorkItem(static s => s.Key(s.Value), new KeyValuePair<SendOrPostCallback, object?>(d, state), preferLocal: false);
```

Reply:
Ok, so it looks like `ThenPost` can be called from any thread. My question was whether it's always called from the scheduler thread. From your answers it looks like it can be called from any thread.